packages/react-native-codegen/src/parsers/typescript/buildSchema.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow strict
+ * @format
+ */
+
+'use strict';
+
+import type {SchemaType} from '../../CodegenSchema';
+import type {Parser} from '../parser';
+
+// $FlowFixMe[untyped-import] Use flow-types for @babel/parser
+const babelParser = require('@babel/parser');
+const {
+  buildSchemaFromConfigType,
+  getConfigType,
+  isModuleRegistryCall,
+} = require('../utils');
+const {buildComponentSchema} = require('./components');
+const {wrapComponentSchema} = require('./components/schema');
+const {buildModuleSchema} = require('./modules');
+
+function Visitor(infoMap: {isComponent: boolean, isModule: boolean}) {
+  return {
+    CallExpression(node: $FlowFixMe) {
+      if (
+        node.callee.type === 'Identifier' &&
+        node.callee.name === 'codegenNativeComponent'
+      ) {
+        infoMap.isComponent = true;
+      }
+
+      if (isModuleRegistryCall(node)) {
+        infoMap.isModule = true;
+      }
+    },
+
+    TSInterfaceDeclaration(node: $FlowFixMe) {
+      if (
+        Array.isArray(node.extends) &&
+        node.extends.some(
+          extension => extension.expression.name === 'TurboModule',
+        )
+      ) {
+        infoMap.isModule = true;
+      }
+    },
+  };
+}
+
+function buildSchema(
+  contents: string,
+  filename: ?string,
+  parser: Parser,
+): SchemaType {
+  // Early return for non-Spec JavaScript files
+  if (
+    !contents.includes('codegenNativeComponent') &&
+    !contents.includes('TurboModule')
+  ) {
+    return {modules: {}};
+  }
+
+  const ast = babelParser.parse(contents, {
+    sourceType: 'module',
+    plugins: ['typescript'],
+  }).program;
```

Comment:
What about creating a function in the `Flow/TypeScriptParser` called `getAST()` that uses the `flow-parser` or the `babelParser` and return an untyped `ast`? 🤔 
If I'm not reading this wrong, that's the only difference between the two `buildSchema` functions. With this change, we should be able to have a common `buildSchema` function that asks to the parser to get the proper AST.

How do you feel about this change?

Reply:
If we go for this path, we can rename the `buildSchema.js` to `Visitor.js` and the two files will contain just the Flow/TypeScript implementation for the visitors. :D

Reply:
that's ok, then!